graftm/hmmer.py

```diff
@@ -568,116 +605,134 @@ def _get_read_names(self, search_result, max_range):
-                                    entry[0] = min(ft)  # ammend the entry if they are
-                                    break  # And break the loop
-                    else:  # if no break occured (no overlap)
+
+        for hit in spans:  # For each of these rows (i.e. hits)
+            i = hit[0]  # set id to i
+            c = hit[1]  # set complement to c
+            ft = [min(hit[2:4]), max(hit[2:4])]  # set span as ft (i.e. from - to) - This is the amount covering the query
+            qs = [min(hit[4:6]), max(hit[4:6])]  # seq the query span to qs - This is the amount covering the HMM
+
+            if ft[0] == ft[1]: continue  # if the span covers none of the query, skip that entry (seen this before)
+
+            if i not in splits:  # If the hit hasnt been seen yet
+                splits[i] = {'span'       : [ft],
+                             'strand'     : [c],
+                             'query_span' : [qs]}  # add the span and complement as new entry
+
+            else:  # otherwise (if it has been seen)
+                for idx, entry in enumerate(splits[i]['span']):  # for each previously added entry
+                    if splits[i]['strand'][idx] != c:  # If the hit is on the same complement strand
                         splits[i]['span'].append(ft)  # Add the new range to be split out in the future
                         splits[i]['strand'].append(c)  # Add the complement strand as well
-    
-        return {key: entry['span'] for key, entry in splits.iteritems()}  # return the dict, without strand information which isn't required.
-    
+                        splits[i]['query_span'].append(qs)
+                        break
+
+                    previous_qs      = splits[i]['query_span'][idx] # Get the query span of the previous hit
+
+                    previous_q_range = set(range(previous_qs[0], previous_qs[1])) # Get the range of each
+                    current_q_range  = set(range(qs[0], qs[1]))
+                    query_overlap    = set(previous_q_range).intersection(current_q_range) # Find the intersection between the two ranges
+
+                    previous_ft_span = set(range(entry[0], entry[1]))
+                    current_ft_span  = set(range(ft[0], ft[1]))
+                    if any(query_overlap): # If there is an overlap
+                        ####################################################
+                        # if the span over the actual read that hit the HMM
+                        # for each hit overlap by > 25%, they are considered
+                        # the same hit, and ignored
+                        ####################################################
+                        intersection_fraction = float(len(previous_ft_span.intersection(current_ft_span)))
+
+                        if intersection_fraction / float(len(previous_ft_span)) >= 0.25:
```

Comment:
0.25 is a constant so should probably be hardcoded at the top of this file or something